File: 섹션06. DFS - 경우의 수 만들기/07-02. BOJ 10819 - 차이를 최대로.cpp
#include <stdio.h>

#define MAX (8 + 3)

int N, M;

int num_of_cases[MAX];
bool visit[MAX];
int number[MAX];

int maxAnswer;

void printCases()
{
	for (int i = 0; i < M; i++) printf("%d ", num_of_cases[i]);
	putchar('\n');
}

int abs(int x)
{
	return (x > 0) ? x : -x;
}

int calculate()
{
	int sum = 0;

	for (int i = 0; i < N - 1; i++)
		sum += abs((num_of_cases[i] - num_of_cases[i + 1]));

	return sum;
}

void DFS(int depth)
{
	if (depth == N)
	{
		// printCases();

		int tmp = calculate();
		if (maxAnswer < tmp) maxAnswer = tmp;

		return;
	}

	for (int i = 0; i < N; i++)
	{
		if (visit[i] == true) continue;

		num_of_cases[depth] = number[i];

		visit[i] = true;
		DFS(depth + 1);
		visit[i] = false;
	}
}

int main()
{
	scanf("%d", &N);

	for (int i = 0; i < N; i++) scanf("%d", &number[i]);

	maxAnswer = -1;

	DFS(0);

	printf("%d\n", maxAnswer);

	return 0;
}

File: Python/섹션06. DFS - 경우의 수 만들기/07-02. BOJ 10819 - 차이를 최대로.py
#C++ 코드, 고치기
MAX = 8 + 3

num_of_cases = [0] * MAX
visit = [False] * MAX
number = [0] * MAX

max_answer = -1

N = int(input())
number[0:N] = list(map(int, input().split()))

def print_cases():
    print(*num_of_cases[:N])

def calculate():
    sum = 0
    
    for i in range(N - 1):
        sum += abs(num_of_cases[i] - num_of_cases[i + 1])
    
    return sum

def DFS(depth):
    global max_answer
    
    if depth == N:
        # print_cases()
        tmp = calculate()
        if max_answer < tmp: max_answer = tmp
        
        return
    
    for i in range(N):
        if visit[i] == True: continue
        
        num_of_cases[depth] = number[i]
        
        visit[i] = True
        DFS(depth + 1)
        visit[i] = False
    
DFS(0)

print(max_answer)